From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Robin Burchell <robin+git@viroteck.net>
Date: Wed, 4 May 2016 08:49:17 -0400
Subject: [PATCH] [maliit] Guard all udev usage with ifndef
 HAVE_CONTEXTSUBSCRIBER. Contributes to JB#34869.

If HAVE_CONTEXTSUBSCRIBER is enabled, the udev codepaths are never used
(detectEvdev is not called). However, we still used udev code in the binary
(udev_ types and functions), which was incorrect because we don't link udev in
this case.

This went undetected until we enabled linking with -Wl,--as-needed, when it
broke.
---
 src/mimhwkeyboardtracker.cpp | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/src/mimhwkeyboardtracker.cpp b/src/mimhwkeyboardtracker.cpp
index 471db3f..b7126b5 100644
--- a/src/mimhwkeyboardtracker.cpp
+++ b/src/mimhwkeyboardtracker.cpp
@@ -69,6 +69,7 @@ MImHwKeyboardTrackerPrivate::MImHwKeyboardTrackerPrivate(MImHwKeyboardTracker *q
 
 void MImHwKeyboardTrackerPrivate::detectEvdev()
 {
+#ifndef HAVE_CONTEXTSUBSCRIBER
     // Use udev to enumerate all input devices, using evdev on each device to
     // find the first device offering a SW_TABLET_MODE switch. If found, this
     // switch is used to determine keyboard presence.
@@ -106,10 +107,12 @@ void MImHwKeyboardTrackerPrivate::detectEvdev()
     }
     udev_enumerate_unref(enumerate);
     udev_unref(udev);
+#endif
 }
 
 void MImHwKeyboardTrackerPrivate::evdevEvent()
 {
+#ifndef HAVE_CONTEXTSUBSCRIBER
     // Parse the evdev event and look for SW_TABLET_MODE status.
 
     struct input_event ev;
@@ -127,11 +130,14 @@ void MImHwKeyboardTrackerPrivate::evdevEvent()
         evdevTabletModePending = -1;
         Q_EMIT stateChanged();
     }
-
+#endif
 }
 
 void MImHwKeyboardTrackerPrivate::tryEvdevDevice(const char *device)
 {
+    Q_UNUSED(device);
+
+#ifndef HAVE_CONTEXTSUBSCRIBER
     QFile *qfile = new QFile(this);
     unsigned char evbits[BITS2BYTES(EV_MAX)];
     int fd;
@@ -185,6 +191,7 @@ void MImHwKeyboardTrackerPrivate::tryEvdevDevice(const char *device)
         return;
 
     evdevTabletMode = TEST_BIT(SW_TABLET_MODE, state);
+#endif
 }
 
 MImHwKeyboardTrackerPrivate::~MImHwKeyboardTrackerPrivate()
-- 
2.35.1

